app/services/training_periods/finish.rb

```diff
@@ -51,5 +51,11 @@ def finish!
         )
       end
     end
+
+  private
+
+    def earliest_finished_on
+      [finished_on, training_period.api_withdrawn_at, training_period.api_deferred_at].compact.min
```

Comment:
ah its here 👍

Reply:
Ah yes, but there's a bug in my spike as this would always be called before the training period being withdrawn/deferred how I've currently got it. I need to switch the order to be:

```
mark_training_period_as_withdrawn!
finish_training_period! if training_period.ongoing?
```